File: conky/playing_clementine.lua
-- Biliothèque de récupération de donées clementine pour conky
-- Ecrit par Didier-T
-- Du forum ubuntu.fr
-- Pour les besoins d'un conky créer par Ferod
-- Le 23/06/2015


local json = require('json')
local posix = require('posix')
local assert = assert
local os = require('os')
local io = require('io')
local dermodif = nil
local info_titre, time_titre
module("playing_clementine")

do
    function info(recherche)
        --Vérifier l'existence d'un fichier
        local function existe(file)
            local a = io.open(file, "r")
            local present
            if a then
                present = true
                io.close(a)
            else
                present = false
            end
            return present
        end
        if existe("/home/sebastien/.conky/data_clem_piste") then
            local dermodif2=posix.stat('/home/sebastien/.conky/data_clem_piste', 'mtime')
            if dermodif == nil or dermodif ~= dermodif2 then
                dermodif=dermodif2
                local file = assert(io.open("/home/sebastien/.conky/data_clem_piste", "rb") )
                local line = file:read()
                info_titre = json.decode(line)
                assert(file:close())
            end
        end
        if existe("/home/sebastien/.conky/data_clem_temps") then
            local dermodif4=posix.stat('/home/sebastien/.conky/data_clem_temps', 'mtime')
            if dermodif3 == nil or dermodif3 ~= dermodif4 then
                dermodif3=dermodif4
                local file = assert(io.open("/home/sebastien/.conky/data_clem_temps", "rb"))
                local line = file:read()
                time_titre = json.decode(line)
                assert(file:close())
            end
        end

        if info_titre==nil or time_titre==nil then
            return
        elseif recherche=="artiste" then
            return info_titre["artist"] or "N/A"
        elseif recherche=="titre" then
            return info_titre["title"] or "N/A"
        elseif recherche=="album" then
            return info_titre["album"] or "N/A"
        elseif recherche=="genre" then
            return info_titre["genre"] or "N/A"
        elseif recherche=="annee" then
            return info_titre["year"] or "N/A"
        elseif recherche=="tracknumber" then
            return info_titre["tracknumber"] or "N/A"
        elseif recherche=="bitrate" then
            return info_titre["bitrate"] or "N/A"
        elseif recherche=="samplerate" then
            return info_titre["samplerate"] or "N/A"
        elseif recherche=="temp_passe" then
            return time_titre["etime"] or "0"
        elseif recherche=="temp_restant" then
            return time_titre["rtime"] or "0"
        elseif recherche=="temp_titre" then
            return time_titre["mtime"] or "0"
        elseif recherche=="avancement" then
            return time_titre["progress"] or "0"
        else
            return "Non affecté"
        end
    end
end

File: conky/playing_clementine.py
#!/usr/bin/env python2
#-*- coding: ascii -*-
#
# playing_clementine.py
# Script creer sur la base de anowplaying.py
# par Didier-T
#
# pour utilisation avec conky
#

import dbus, optparse, shutil, commands
import unicodedata
import math
from time import sleep


class maj:
    def __init__(self):
        '''Get system bus'''
        bus = dbus.SessionBus()
        try:
            self.amarok = bus.get_object('org.mpris.MediaPlayer2.clementine', '/org/mpris/MediaPlayer2')
        except:
            self.amarok=None
            return
        self.amarokdict = self.amarok.Get('org.mpris.MediaPlayer2.Player', 'Metadata', dbus_interface='org.freedesktop.DBus.Properties')
        self.holdtitle=""
        self.holdcover=""

    def unaccent(self, str):
        return unicodedata.normalize('NFKD', str).encode('ascii', 'ignore')

    def init_ok(self):
        if self.amarok is not None:
            return 0
        else:
            return 1

    def maj(self):
        temp = {}
        ret = {}
        cpos = mt = mtime = etime = rtime = progress = None

        if self.amarokdict.has_key('mpris:length'):
            try :
                cpos = self.amarok.Get('org.mpris.MediaPlayer2.Player', 'Position', dbus_interface='org.freedesktop.DBus.Properties')/1000000
            except:
                self.amarok=None
                return

            mt = self.amarokdict['mpris:length']/1000000
            mtime = str(mt/60)+":"+str(mt%60) if mt%60>9 else str(mt/60)+":0"+str(mt%60)
            etime = str(cpos/60)+":"+str(cpos%60) if cpos%60>9 else str(cpos/60)+":0"+str(cpos%60)
            rtime = str((mt-cpos)/60)+":"+str((mt-cpos)%60) if (mt-cpos)%60>9 else str((mt-cpos)/60)+":0"+str((mt-cpos)%60)
            progress= int(float(cpos)/float(mt)*100)

        if etime is not None:
            temp["etime"]=etime
        if rtime is not None:
            temp["rtime"]=rtime
        if mtime is not None:
            temp["mtime"]=mtime
        if progress is not None:
            temp["progress"]=progress

        try :
            self.amarokdict = self.amarok.Get('org.mpris.MediaPlayer2.Player', 'Metadata', dbus_interface='org.freedesktop.DBus.Properties')
        except:
            self.amarok=None
            return

        if self.amarokdict.has_key('xesam:artist'):
            ret["artist"] = self.unaccent(self.amarokdict['xesam:artist'][0][0:40])
        if self.amarokdict.has_key('xesam:title'):
            ret["title"] = self.unaccent(self.amarokdict['xesam:title'][0:40])
        else :
            ret["title"] = self.holdtitle
        if self.amarokdict.has_key('xesam:album'):
            ret["album"] = self.unaccent(self.amarokdict['xesam:album'][0:35])
        if self.amarokdict.has_key('xesam:genre'):
            ret["genre"] = self.unaccent(self.amarokdict['xesam:genre'][0][0:40])
        if self.amarokdict.has_key('year'):
            ret["year"] = str(self.amarokdict['year'])
        if self.amarokdict.has_key('xesam:trackNumber'):
            ret["tracknumber"] = str(self.amarokdict['xesam:trackNumber'])
        if self.amarokdict.has_key('bitrate'):
            ret["bitrate"] = str(self.amarokdict['bitrate'])
        if self.amarokdict.has_key('audio-samplerate'):
            ret["samplerate"] = str(self.amarokdict['audio-samplerate'])

        if self.amarokdict.has_key('mpris:artUrl'):
            cover = self.amarokdict['mpris:artUrl']
            if cover != "" and self.holdcover != cover :
                self.holdcover=cover
                try :
                    shutil.copyfile(cover.replace('file://', ''), "/home/sebastien/.conky/jaquet.jpg")
                except Exception, e:
                    print e

        if self.holdtitle != ret["title"] :
            self.holdtitle = ret["title"]
            fiche = open("/home/sebastien/.conky/data_clem_piste", "w")
            fiche.write(str(ret))
            fiche.close()

        fiche = open("/home/sebastien/.conky/data_clem_temps", "w")
        fiche.write(str(temp))
        fiche.close()


def demonise(init):
    x=0
    while 1 :
        '''Check if clementine is running'''
        output = commands.getoutput('ps -A')
        if 'clementine' in output:
            if init==1 :
                miseajour=maj()
                init=miseajour.init_ok()
            else:
                init=miseajour.init_ok()
                if init==0:
                    miseajour.maj()

        sleep(1)


if __name__ == '__main__':
    demonise(1)
